crates/driver/src/domain/competition/solution/scoring.rs

```diff
@@ -0,0 +1,262 @@
+    custom_price: CustomClearingPrices,
+    policies: Vec<order::FeePolicy>,
+}
+
+impl Trade {
+    pub fn new(
+        sell: eth::Asset,
+        buy: eth::Asset,
+        side: Side,
+        executed: order::TargetAmount,
+        custom_price: CustomClearingPrices,
+        policies: Vec<order::FeePolicy>,
+    ) -> Self {
+        Self {
+            sell,
+            buy,
+            side,
+            executed,
+            custom_price,
+            policies,
+        }
+    }
+
+    /// CIP38 score defined as surplus + protocol fee
+    ///
+    /// Denominated in NATIVE token
+    fn score(&self, prices: &auction::Prices) -> Result<eth::TokenAmount, Error> {
+        Ok(self.native_surplus(prices)? + self.native_protocol_fee(prices)?)
+    }
+
+    /// Surplus based on custom clearing prices returns the surplus after all
+    /// fees have been applied.
+    ///
+    /// Denominated in SURPLUS token
+    fn surplus(&self) -> Option<eth::Asset> {
+        match self.side {
+            Side::Buy => {
+                // scale limit sell to support partially fillable orders
+                let limit_sell = self
+                    .sell
+                    .amount
+                    .0
+                    .checked_mul(self.executed.into())?
+                    .checked_div(self.buy.amount.into())?;
+                // difference between limit sell and executed amount converted to sell token
+                limit_sell.checked_sub(
+                    self.executed
+                        .0
+                        .checked_mul(self.custom_price.buy)?
+                        .checked_div(self.custom_price.sell)?,
+                )
+            }
+            Side::Sell => {
+                // scale limit buy to support partially fillable orders
+                let limit_buy = self
+                    .executed
+                    .0
+                    .checked_mul(self.buy.amount.into())?
+                    .checked_div(self.sell.amount.into())?;
+                // difference between executed amount converted to buy token and limit buy
+                self.executed
+                    .0
+                    .checked_mul(self.custom_price.sell)?
+                    .checked_div(self.custom_price.buy)?
+                    .checked_sub(limit_buy)
+            }
+        }
+        .map(|surplus| eth::Asset {
+            token: self.surplus_token(),
+            amount: surplus.into(),
+        })
+    }
+
+    /// Surplus based on custom clearing prices returns the surplus after all
+    /// fees have been applied.
+    ///
+    /// Denominated in NATIVE token
+    fn native_surplus(&self, prices: &auction::Prices) -> Result<eth::TokenAmount, Error> {
+        let surplus = self.surplus_token_price(prices)?.apply(
+            self.surplus()
+                .ok_or(Error::Surplus(self.sell, self.buy))?
+                .amount,
+        );
+        // normalize
+        Ok((surplus.0 / *UNIT).into())
```

Comment:
Note, there is `Prices::ether_value` that can be used here:

https://github.com/cowprotocol/services/blob/7304593510eea3e32a2e7097b06543f38f203aeb/crates/solvers/src/domain/auction.rs#L78-L82